frontend/controller/serviceworkers/push.js

```diff
@@ -7,28 +7,83 @@ import { NOTIFICATION_TYPE, PUBSUB_RECONNECTION_SUCCEEDED, PUSH_SERVER_ACTION_TY
 import { getSubscriptionId } from '~/shared/functions.js'
 import { DEVICE_SETTINGS } from '@utils/constants.js'
 
+// The application server (public) key could be either an ArrayBuffer (which is
+// what we get from fetching the current subscription), or it could be a
+// base64url-encoded string (which is what we usually get from the server)
+// This string coerces both of these into an `Uint8Array` instance.
+const strOrBufToBuf = (v: string | Uint8Array | ArrayBuffer) => {
+  if (typeof v === 'string') {
+    v = Buffer.from(
+      // Convert from base64url to base64
+      v.replace(/_/g, '/').replace(/-/g, '+') + '='.repeat((4 - v.length % 4) % 4),
+      'base64'
+    )
+    return v
+  }
+  return new Uint8Array(v)
+}
+
+const bufferEq = (a?: ArrayBuffer | Uint8Array, b?: ArrayBuffer | Uint8Array) => {
+  // eslint-disable-next-line eqeqeq
+  if (a == null || b == null) return a == b
+
+  const ab = strOrBufToBuf(a)
+  const bb = strOrBufToBuf(b)
+
+  if (ab.byteLength !== bb.byteLength) return false
+
+  for (let i = ab.byteLength - 1; i >= 0; i--) {
+    if (ab[i] !== bb[i]) return false
+  }
+
+  return true
+}
+
 export default (sbp('sbp/selectors/register', {
   'push/getSubscriptionOptions': (() => {
     let cachedVapidInformation
+
+    sbp('okTurtles.events/on', REQUEST_TYPE.PUSH_ACTION, ({ data }) => {
+      if (data.type !== PUSH_SERVER_ACTION_TYPE.SEND_PUBLIC_KEY) return
+      const oldKey = cachedVapidInformation?.[1].applicationServerKey
+      cachedVapidInformation = [performance.now(), {
+        userVisibleOnly: true,
+        applicationServerKey: data.data
+      }]
+      if (oldKey === data.data) return
+      (async () => {
+        const subscription = await self.registration.pushManager.getSubscription()
+        if (!subscription) return
+        if (bufferEq(subscription.options.applicationServerKey, data.data)) return
+        console.warn('VAPID server key changed; removing existing subscription and setting up a new one', {
+          oldApplicationServerPublicKey: subscription.options.applicationServerKey && Buffer.from(subscription.options.applicationServerKey).toString('base64'),
+          newApplicationServerPublicKey: data.data
+        })
+        await subscription.unsubscribe()
```

Comment:
Should we not proceed even if this throws? 🤔 

Seems like we should still try to proceed if this throws...

Reply:
I don't think so here. Proceed and do what, exactly?

We call unsubscribe to remove the existing subscription and then the rest of the logic is to create a new subscription. If unsubscribing fails, we'll get an error like `DOMException: A subscription with a different application server key already exists.`

Since the rest of the code only makes sense after unsubscribing, then I'd say this is fine as is.